File: backend/history_manager.py
import json
import os
from datetime import datetime
from typing import List, Dict, Optional
import uuid

class HistoryManager:
    """Manages evaluation history using JSON file storage"""

    def __init__(self, history_file: str = None):
        # Use /tmp for Vercel/serverless environments (only writable directory)
        if history_file is None:
            # Check if running in Vercel/serverless environment
            # Vercel sets VERCEL env var, Lambda/AWS sets AWS_LAMBDA_FUNCTION_VERSION
            is_serverless = (
                os.environ.get('VERCEL') is not None
                or os.environ.get('AWS_LAMBDA_FUNCTION_VERSION') is not None
                or os.path.exists('/var/task')  # Vercel/AWS Lambda marker
            )
            if is_serverless:
                history_file = "/tmp/evaluations_history.json"
            else:
                history_file = "evaluations_history.json"
        self.history_file = history_file
        self._ensure_file_exists()
    
    def _ensure_file_exists(self):
        """Create history file if it doesn't exist"""
        if not os.path.exists(self.history_file):
            with open(self.history_file, 'w') as f:
                json.dump([], f)
    
    def _read_history(self) -> List[Dict]:
        """Read all history records"""
        try:
            with open(self.history_file, 'r', encoding='utf-8') as f:
                return json.load(f)
        except Exception as e:
            print(f"Error reading history: {e}")
            return []
    
    def _write_history(self, history: List[Dict]):
        """Write history records to file"""
        try:
            with open(self.history_file, 'w', encoding='utf-8') as f:
                json.dump(history, f, ensure_ascii=False, indent=2)
        except Exception as e:
            print(f"Error writing history: {e}")
    
    def save_evaluation(self, 
                       report: Dict,
                       teacher_doc_name: str,
                       dialogue_record_name: str,
                       model: str = "gpt-4o") -> str:
        """
        Save an evaluation to history
        Returns the generated history ID
        """
        history = self._read_history()
        
        # Generate unique ID
        eval_id = str(uuid.uuid4())
        
        # Create history entry
        entry = {
            "id": eval_id,
            "timestamp": datetime.now().isoformat(),
            "teacher_doc_name": teacher_doc_name,
            "dialogue_record_name": dialogue_record_name,
            "model": model,
            "total_score": report.get("total_score", 0),
            "final_level": report.get("final_level", ""),
            "report": report  # Full report data
        }
        
        # Add to beginning of list (most recent first)
        history.insert(0, entry)
        
        # Limit history to last 100 entries
        if len(history) > 100:
            history = history[:100]
        
        self._write_history(history)
        return eval_id
    
    def get_all(self) -> List[Dict]:
        """Get all evaluation history (summary only, without full reports)"""
        history = self._read_history()
        # Return summaries without full report data to reduce payload
        return [{
            "id": entry["id"],
            "timestamp": entry["timestamp"],
            "teacher_doc_name": entry["teacher_doc_name"],
            "dialogue_record_name": entry["dialogue_record_name"],
            "model": entry["model"],
            "total_score": entry["total_score"],
            "final_level": entry["final_level"]
        } for entry in history]
    
    def get_by_id(self, eval_id: str) -> Optional[Dict]:
        """Get a specific evaluation by ID"""
        history = self._read_history()
        for entry in history:
            if entry["id"] == eval_id:
                return entry
        return None
    
    def delete_by_id(self, eval_id: str) -> bool:
        """Delete an evaluation by ID. Returns True if successful."""
        history = self._read_history()
        original_length = len(history)
        history = [entry for entry in history if entry["id"] != eval_id]
        
        if len(history) < original_length:
            self._write_history(history)
            return True
        return False

File: frontend/lib/history-manager.ts
/**
 * 历史记录管理器
 * 使用 JSON 文件存储评估历史
 */

import { promises as fs } from 'fs';
import path from 'path';
import { v4 as uuidv4 } from 'uuid';

export interface HistoryEntry {
    id: string;
    timestamp: string;
    teacher_doc_name: string;
    dialogue_record_name: string;
    model: string;
    total_score: number;
    final_level: string;
    report: Record<string, unknown>;
}

export interface HistorySummary {
    id: string;
    timestamp: string;
    teacher_doc_name: string;
    dialogue_record_name: string;
    model: string;
    total_score: number;
    final_level: string;
}

const HISTORY_FILE = process.env.HISTORY_FILE || 'evaluations_history.json';

async function getHistoryFilePath(): Promise<string> {
    // In production, use a data directory
    const dataDir = process.env.DATA_DIR || process.cwd();
    return path.join(dataDir, HISTORY_FILE);
}

async function ensureFileExists(): Promise<void> {
    const filePath = await getHistoryFilePath();
    try {
        await fs.access(filePath);
    } catch {
        await fs.writeFile(filePath, '[]', 'utf-8');
    }
}

async function readHistory(): Promise<HistoryEntry[]> {
    await ensureFileExists();
    const filePath = await getHistoryFilePath();
    try {
        const content = await fs.readFile(filePath, 'utf-8');
        return JSON.parse(content);
    } catch (error) {
        console.error('Error reading history:', error);
        return [];
    }
}

async function writeHistory(history: HistoryEntry[]): Promise<void> {
    const filePath = await getHistoryFilePath();
    try {
        await fs.writeFile(filePath, JSON.stringify(history, null, 2), 'utf-8');
    } catch (error) {
        console.error('Error writing history:', error);
    }
}

/**
 * 保存评估到历史记录
 */
export async function saveEvaluation(
    report: Record<string, unknown>,
    teacherDocName: string,
    dialogueRecordName: string,
    model: string = 'gpt-4o'
): Promise<string> {
    const history = await readHistory();

    const evalId = uuidv4();

    const entry: HistoryEntry = {
        id: evalId,
        timestamp: new Date().toISOString(),
        teacher_doc_name: teacherDocName,
        dialogue_record_name: dialogueRecordName,
        model,
        total_score: (report.total_score as number) || 0,
        final_level: (report.final_level as string) || '',
        report
    };

    // 添加到列表开头（最新的在前）
    history.unshift(entry);

    // 限制历史记录数量
    if (history.length > 100) {
        history.length = 100;
    }

    await writeHistory(history);
    return evalId;
}

/**
 * 获取所有历史记录摘要
 */
export async function getAllHistory(): Promise<HistorySummary[]> {
    const history = await readHistory();
    return history.map(entry => ({
        id: entry.id,
        timestamp: entry.timestamp,
        teacher_doc_name: entry.teacher_doc_name,
        dialogue_record_name: entry.dialogue_record_name,
        model: entry.model,
        total_score: entry.total_score,
        final_level: entry.final_level
    }));
}

/**
 * 根据 ID 获取特定评估
 */
export async function getHistoryById(evalId: string): Promise<HistoryEntry | null> {
    const history = await readHistory();
    return history.find(entry => entry.id === evalId) || null;
}

/**
 * 删除指定 ID 的评估
 */
export async function deleteHistoryById(evalId: string): Promise<boolean> {
    const history = await readHistory();
    const originalLength = history.length;
    const filtered = history.filter(entry => entry.id !== evalId);

    if (filtered.length < originalLength) {
        await writeHistory(filtered);
        return true;
    }
    return false;
}
